File: OsmAndWear/src/net/osmand/plus/wikivoyage/data/PopularArticles.java
package net.osmand.plus.wikivoyage.data;

import androidx.annotation.NonNull;

import net.osmand.util.Algorithms;

import java.util.ArrayList;
import java.util.List;

public class PopularArticles {

	public static final int ARTICLES_PER_PAGE = 30;

	private List<TravelArticle> articles;

	public PopularArticles() {
		this.articles = new ArrayList<>();
	}

	public PopularArticles(@NonNull PopularArticles articles) {
		this.articles = articles.articles;
	}

	public void clear() {
		articles = new ArrayList<>();
	}

	@NonNull
	public List<TravelArticle> getArticles() {
		return new ArrayList<>(articles);
	}

	public boolean add(@NonNull TravelArticle article) {
		articles.add(article);
		return articles.size() % ARTICLES_PER_PAGE != 0;
	}

	public boolean contains(@NonNull TravelArticle article) {
		return articles.contains(article);
	}

	public boolean containsByRouteId(@NonNull String routeId) {
		for (TravelArticle article : articles) {
			if (article.getRouteId().equals(routeId)) {
				return true;
			}
		}
		return false;
	}

	public boolean isEmpty() {
		return Algorithms.isEmpty(articles);
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/travel/PopularArticles.kt
package net.osmand.shared.travel

import net.osmand.shared.util.KAlgorithms

/**
 * One page of the articles offered on the travel screen before anything is searched, filled a page
 * at a time as the search widens.
 *
 * A copy of `PopularArticles` in the android app, which stays there; this copy is for iOS. Note
 * that the copy constructor shares the list with the original rather than copying it, as java
 * does: that is how a page carries over what the previous one collected.
 */
class PopularArticles {

	private var articles: MutableList<TravelArticle>

	constructor() {
		articles = ArrayList()
	}

	constructor(other: PopularArticles) {
		articles = other.articles
	}

	fun clear() {
		articles = ArrayList()
	}

	fun getArticles(): List<TravelArticle> = ArrayList(articles)

	/** @return false once the page is full. */
	fun add(article: TravelArticle): Boolean {
		articles.add(article)
		return articles.size % ARTICLES_PER_PAGE != 0
	}

	fun contains(article: TravelArticle): Boolean = articles.contains(article)

	fun containsByRouteId(routeId: String): Boolean {
		for (article in articles) {
			if (article.getRouteId() == routeId) {
				return true
			}
		}
		return false
	}

	fun isEmpty(): Boolean = KAlgorithms.isEmpty(articles)

	companion object {
		const val ARTICLES_PER_PAGE: Int = 30
	}
}
